extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,361 @@
+		}, properties));
+	}
+
+	public async getChildren(): Promise<ControllerNode[]> {
+		return this.children as ControllerNode[];
+	}
+
+	public addControllerNode(url: string, username: string, password: string, rememberPassword: boolean, masterInstance?: IEndPoint): void {
+		let controllerNode = this.getExistingControllerNode(url, username);
+		if (controllerNode) {
+			controllerNode.password = password;
+			controllerNode.rememberPassword = rememberPassword;
+			controllerNode.clearChildren();
+		} else {
+			controllerNode = new ControllerNode({ url, username, password, rememberPassword, parent: this, treeChangeHandler: this.treeChangeHandler });
+			this.addChild(controllerNode);
+		}
+
+		if (masterInstance) {
+			controllerNode.addSqlMasterNode(masterInstance.endpoint, masterInstance.description);
+		}
+	}
+
+	public deleteControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		let index = nodes.findIndex(e => e.url === url && e.username === username);
+		let deleted = undefined;
+		if (index >= 0) {
+			deleted = nodes.splice(index, 1);
+		}
+		return deleted;
+	}
+
+	private getExistingControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		return nodes.find(e => e.url === url && e.username === username);
+	}
+}
+
+export class ControllerNode extends ControllerTreeNode {
+	private _url: string;
+	private _username: string;
+	private _password: string;
+	private _rememberPassword: boolean;
+
+	constructor(properties?: {
+		url: string,
+		username: string,
+		password: string,
+		parent: ControllerTreeNode,
+		treeChangeHandler: IControllerTreeChangeHandler,
+		label?: string,
+		description?: string,
+		rememberPassword?: boolean
+	}) {
+		super(Object.assign({
+			label: undefined,
+			nodeType: BdcItemType.controller,
+			iconPath: IconPath.controllerNode
+		}, properties));
+
+		let address = ControllerNode.toIpAndPort(properties.url);
+		this.label = properties.label || `controller: ${address} (${properties.username})`;
+		this.description = properties.description || this.label;
+		this._url = properties.url;
+		this._username = properties.username;
+		this._password = properties.password;
+		this._rememberPassword = !!properties.rememberPassword;
+	}
+
+
+	public async getChildren(): Promise<ControllerTreeNode[]> {
+		if (this.children && this.children.length > 0) {
+			this.clearChildren();
+		}
+
+		if (!this._password) {
+			vscode.commands.executeCommand('bigDataClusters.command.addController', this);
+			return this.children as ControllerTreeNode[];
+		}
+
+		try {
+			let response = await BdcController.getEndPoints(this._url, this._username, this._password, true);
+			if (response && response.endPoints) {
+				let master = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
```

Comment:
>'sql-server-master' [](start = 67, length = 19)

This is used in a couple of places. Anything like this that's used like that should really be defined as a constant somewhere - even if it's just the constants.ts file

Reply:
It is being used 2 times. I will consider making it as const. Using the same string, if it occurs not frequently, happens lots of times in the other extensions though.